### ISSUE TITLE
A way to specify getter/setter logic.

### ISSUE THREAD

**Issue by trusktr** [skatejs/skatejs#1385]
It'd be nice to have a way to implement custom prop handling. I do this in my non-Skate elements with custom getters/setters. It's not possible with Skate because Skate defines the getters/setters.

## Problem:

There's the `coerce` API, but that is limited because `this` can not be accessed inside the `coerce` function to work with the instance that is receiving a property value.

## example of solution in a non-SkateJS way:

First, let me explain what [these](https://github.com/trusktr/infamous/) custom elements of mine do. In those elements, I'm using `attributeChangedCallback` directly and manually creating accessors so that I can do the following:

```js
el.rotation = {x:30, z:45}
```

What this actually does behind the scenes is takes those values, and applies them to a private object that has event listeners and other useful features that I want to preserve across setter invocations. F.e. a simple summary implementation of the above `rotation` setter may look like this:

```
set rotation(val) {
  if (typeof val.x != 'undefined') this._rotation.x = val.x
  if (typeof val.y != 'undefined') this._rotation.y = val.y
  if (typeof val.z != 'undefined') this._rotation.z = val.z
  // Updating `this._rotation.x` causes an event to be fired, etc.
}
```

I do the same thing when processing attributes (f.e. serialize into an object, then set values with the setters).

It'd be great to have a way to achieve this with a SkateJS-based components, but due to the fact that SkateJS defines the setters/getters, there isn't a non-monkey-patch way to do it.

## A solution?

Maybe there can be some way to configure the setup of the props cache, as well as defining get/set behavior somehow. I'm not really sure what the API would be like, but in the end, I'd like to do as little as possible just like how currently we can define attributes that map to properties without having to wire it up.

Even in the latest v6-branch API (at time of writing this), the call to `propertyFunc(newValue)` doesn't seem to provide a way to work with `this`, but maybe it doesn't really make sense to work with it there.

Maybe, we need a way to define _where_ values get saved to. Right now (`v6`), values get saved to Symbols (seems cool) by default, and maybe that's a good default. Maybe there can be some way to configure this?

It's hard to imagine a generic and clean solution...

**Comment by treshugart** [comment]
Why not just set the prop: `rotation: props.object`, and then override the setter `set rotation(obj)` and do your thing there?

**Comment by trusktr** [comment]
> Why not just set the prop: `rotation: props.object`, and then override the setter `set rotation(obj)` and do your thing there?

It also doesn't seem possible to do it inside a `class` definition because the SkateJS `observedAttribute` static getter will run after the class has already been defined when passed to `customElements.define`, so it will override the user's defined getters/setters.

Plus, the overriding developer would have to make sure to patch the SkateJS functionality back in, which is hacky.

Maybe, conversely, SkateJS can patch the end-developer's getters/setters in, so

## idea

After thinking about the [Symbols functionality](https://github.com/skatejs/skatejs/blob/4c3212d7b8b70cff187b80d6751b41dc3f98daa6/packages/skatejs/src/util.js#L23-L29), maybe it can be used to allow getters/setters to be defined for props.

This would be advantageous because `this` inside the getters/setters would be the instance of the class, unlike if getters/setters were placed into a `_props` sub-object.

The functionality that we'd be taking advantage of is this (try it in console):

```js
var f = Symbol('f')
var o = {
  get [f](){ return this._val },
  set [f](val) { this._val = val },
}
o[f] = 5
o[f] // 5
```

## Possible API

It could possibly look like follows, with the advantage being that the getters/setters generated by SkateJS would still be able to do their thing:

```js
class MyEl extends WithUpdate(HTMLElement) {
    static get props() {
        return {
            rotation: {
                ...props.object,
                get: function() {
                    return this._val
                },
                set(val) {
                    this._val = val
                }
            }
        }
    }
}
```

## implementation idea 1

The [relevant part](https://github.com/skatejs/skatejs/blob/4c3212d7b8b70cff187b80d6751b41dc3f98daa6/packages/skatejs/src/with-update.js#L89-L109) of `with-update.js` would look like this:

```js
    Object.defineProperty(constructor.prototype, name, {
      configurable: true,
      get() {
        const val = normalised.get.call(this); // <--- RIGHT HERE
        return val == null ? normalised.default : val;
      },
      set(val) {
        const { attribute: { target }, serialize } = normalised;
        if (target) {
          const serializedVal = serialize ? serialize(val) : val;
          if (serializedVal == null) {
            this.removeAttribute(target);
          } else {
            this.setAttribute(target, serializedVal);
          }
        }
        normalised.set.call(this, normalised.coerce(val)); // <--- AND HERE
        this.triggerUpdate();
      }
    });
  };
```

A downside to this is that getters/setters would be defined inside the static `props` getter, which isn't as clean and semantic as defining them inside the class.

## implementation idea 2

Maybe the [`defineProps(this)` call](https://github.com/skatejs/skatejs/blob/4c3212d7b8b70cff187b80d6751b41dc3f98daa6/packages/skatejs/src/with-update.js#L154) could first scan the class prototype, save the class' defined getters/setters, and then do the same thing as above (somewhat like a decorator would modify a class).

I tested to be sure, and `super` works as expected:

```js
{
class Bar {
 get foo() {return this._foo}
}
class Foo extends Bar {
 get foo() {return super.foo}
 set foo(v) {this._foo = v}
}

const { get:getFoo, set:setFoo } = Object.getOwnPropertyDescriptor(Foo.prototype, 'foo')

Object.defineProperty(Foo.prototype, 'foo', {
 configurable: true,
 get() {
  console.log('get foo')
  return getFoo.call(this)
 },
 set(v) {
  console.log('set foo')
  setFoo.call(this, v)
 }
})

const foo = new Foo

foo.foo = 5 // outputs "set foo"
foo.foo // outputs "get foo"
}
```

**Comment by trusktr** [comment]
Implementation idea 2 is better, because `super` doesn't work in idea 1 and requires pre-ES6 verbosity to make super-like functionality.

**Comment by trusktr** [comment]
`normalised.set` can default to setting `this._props[name] = val` in the default case, and similar with `get`.

**Comment by trusktr** [comment]
Example usage of implementation idea 2:

```js
import { props } from 'skatejs'

class MyEl extends withUpdate(HTMLElement) {
  static props = {
    rotation: props.object,
  }

  get rotation() { /* ... custom getter logic, then wrapped by SkateJS ... */ }
  set rotation(val) { /* ... custom setter logic, then wrapped by SkateJS ... */ }
}
```

Otherwise, if the user leaves out the getter/setter, 

```js
import { props } from 'skatejs'

class MyEl extends withUpdate(HTMLElement) {
  static props = {
    rotation: props.object,
  }
}
```

then SkateJS can normalize to using `this._props` like before.

**Comment by treshugart** [comment]
Option 2 is an interesting idea. Similar to what you've done, you can basically do this internally:

```js
Object.defineProperty(proto, propName, {
  configurable: true,
  ...propDefinition,
  ...Object.getOwnPropertyDescriptor(proto, propName)
})
```

**Comment by trusktr** [comment]
> Similar to what you've done, you can basically do this internally

Yes, but that doesn't let Skate be able to do custom things during get/set, which is what I was trying to preserve. F.e. Skate was removing an attribute if it found the value to be `null`.

**Comment by treshugart** [comment]
> Yes, but that doesn't let Skate be able to do custom things during get/set, which is what I was trying to preserve. F.e. Skate was removing an attribute if it found the value to be null.

Wouldn't `...Object.getOwnPropertyDescriptor(proto, propName)` do that? Maybe I missed something.

**Comment by trusktr** [comment]
That will override the get/set definitions that are defined in `...propDefinition` right? So those ones defined by Skate will not do anything special because they are replaced by the user's.

**Comment by treshugart** [comment]
Yeah, I think that's the idea. Unless it's replaced with `null` or something, then that'd be unwanted. There'd have to be an order of precedence no matter what.

**Comment by trusktr** [comment]
> Yeah, I think that's the idea.

But it will break how SkateJS works. Anywho, I think the use case for this is basically covered by the PR I linked in #1493.

I experimented with implementing this, but I found it to be redundant when paired with the changes in https://github.com/skatejs/skatejs/pull/1509, which allows me to cover the use case.

Basically, `coerce` is called in the same place we'd call a `set` function, which I found to be redundant if `coerce` can already access `this`. The withUpdate getter will already be able to return the cache location that is used in `coerce` as I've shown in #1509.

**Comment by treshugart** [comment]
Just for posterity, `v6-trey` now has a `changed` callback (does not change the value, but is called post change). This should be able to satisfy this use case.

**Comment by trusktr** [comment]
I would love to try our the new `get`/`set`, etc, but at the moment I can't because I am heavily relying on `withUpdate` being a mixin with no DOM APIs.

**Comment by treshugart** [comment]
> I'm using withUpdate on both Custom Elements as well as non-DOM objects.

Can you open a separate issue to discuss this specifically? It's definitely not meant to be used this way.
